Create privatelink-with-network-firewall.yaml

niroy opened the pull request:
Modified the rosa-privatelink cloudformation template to include network firewall. 
1. Some paths with / are not accepted such as cloud.redhat.com/api/ingress
2. Few domains need port 22 and port 9997 , those have not been implemented (not sure how)

niroy committed:
Create privatelink-with-network-firewall.yaml

Modified the rosa-privatelink cloudformation template to include network firewall. 
1. Some paths with / are not accepted such as cloud.redhat.com/api/ingress
2. Few domains need port 22 and port 9997 , those have not been implemented (not sure how)

herve4m committed:
Add VPC default network ACL to the template

herve4m commented on privatelink-with-network-firewall.yaml:
```diff
@@ -0,0 +1,746 @@
+      Tags:
+        - Key: "Name"
+          Value: !Join ["", [!Ref pRosaClusterName, "-firewall-rule-group"]]
+      RuleGroup:
```
I think you also need to set the `HOME_NET` variable to filter traffic from the ROSA VPC.
See https://docs.aws.amazon.com/network-firewall/latest/developerguide/stateful-rule-groups-domain-names.html#stateful-rule-groups-domain-names-home-net

```suggestion
      RuleGroup:
        RuleVariables:
          IPSets:
            HOME_NET:
              Definition:
                - "10.0.0.0/16"
                - "10.1.0.0/16"
```

niroy replied:
Thanks, added

herve4m commented on privatelink-with-network-firewall.yaml:
```diff
@@ -0,0 +1,746 @@
+          - "aws:forward_to_sfe"
+        StatefulRuleGroupReferences:
+          - ResourceArn: !Ref rEgressFirewallRuleGroup
+      Description: Stateful rule group for Network Firewall
```
I would also add a tag. Maybe:

```suggestion
      Description: Stateful rule group for Network Firewall
      Tags:
        - Key: "Name"
          Value: !Join ["", [!Ref pRosaClusterName, "-network-firewall-policy"]]
```

niroy replied:
Added tag

herve4m commented:
1. (filter by path), you have to allow the hostname (cloud.redhat.com): you cannot filter by path because it's 443 (TLS) so the firewall cannot inspect the URL.

2. For port 9997, it seems that the URLs (inputs1.osdsecuritylogs.splunkcloud.com for example) use TLS (but not HTTPS):

```
$ openssl s_client -connect inputs1.osdsecuritylogs.splunkcloud.com:9997
CONNECTED(00000003)
depth=2 CN = Root CA
verify error:num=19:self-signed certificate in certificate chain
verify return:1
depth=2 CN = Root CA
verify return:1
depth=1 CN = Intermediate CA
verify return:1
depth=0 CN = *.osdsecuritylogs.splunkcloud.com
verify return:1
804B7223507F0000:error:0A000410:SSL routines:ssl3_read_bytes:sslv3 alert handshake failure:ssl/record/rec_layer_s3.c:1605:SSL alert number 40
---
Certificate chain
 0 s:CN = *.osdsecuritylogs.splunkcloud.com
   i:CN = Intermediate CA
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Oct  3 01:11:44 2022 GMT; NotAfter: Oct  2 01:12:14 2027 GMT
 1 s:CN = Intermediate CA
   i:CN = Root CA
   a:PKEY: rsaEncryption, 2048 (bit); sigalg: RSA-SHA256
   v:NotBefore: Oct  3 22:50:18 2019 GMT; NotAfter: Sep 30 22:50:48 2029 GMT
 2 s:CN = Root CA
   i:CN = Root CA
   a:PKEY: rsaEncryption, 4096 (bit); sigalg: RSA-SHA256
   v:NotBefore: Oct  3 04:58:37 2019 GMT; NotAfter: Sep 30 04:59:01 2029 GMT
---
...
```

You have to use suricata rules (called signatures) in a AWS::NetworkFirewall::RuleGroup > Properties > RuleGroup > RulesSource > RulesString.
Suricata doc: https://docs.suricata.io/en/latest/rules/index.html
For that case, use the `tls.cert_subject` filter. Something such as:

```
pass tls $HOME_NET any -> $EXTERNAL_NET 9997 (tls.cert_subject; dotprefix; content:".osdsecuritylogs.splunkcloud.com"; endswith; msg:"Pass TLS to .osdsecuritylogs.splunkcloud.com"; sid:1004;  rev:1;)
```

Example on adding such a rule in CloudFormation: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-networkfirewall-rulegroup.html#aws-resource-networkfirewall-rulegroup--examples


For port 22, get the IP address of sftp.access.redhat.com and filter (suricata rule or stateless rule) by IP.

herve4m commented:
You also need to connect the firewall. I think you need to add another private subnet that hosts the firewall (between rEgressVpcPrivateSubnet and rEgressVpcPublicSubnet), add a route table, and update the other route tables.
Example: https://medium.com/slalom-technology/secure-internet-access-egress-filtering-with-aws-network-firewall-ddf52ae121f9

niroy committed:
Add two more firewall security rule groups

Changes include:
1. Adding firewall security rule groups for sftp and osdlogs
2. Adding missing tags
3. Add source IP to rule groups

niroy committed:
Update quotes for uniformity

herve4m commented on privatelink-with-network-firewall.yaml:
```diff
@@ -0,0 +1,818 @@
+              - "HTTP_HOST"
+            Targets:
+              - "registry.redhat.io"
+              - ".quay.io"
```
Missing `quay.io` (in addition to `.quay.io`):
```suggestion
              - ".quay.io"
              - "quay.io"
```

niroy replied:
This gives error if I include quay.io 
`Resource handler returned message: "Targets has duplicate values, parameter: [quay.io], context: RulesSource.RulesSourceList.Targets[2] (Service: NetworkFirewall, Status Code: 400, Request ID: 8dabff27-e8e4-40fa-aba4-e5928bddc0d0)" (RequestToken: b189b206-f146-e004-56a9-1a9d75ac1dc9, HandlerErrorCode: InvalidRequest)`

herve4m replied:
Fixed

niroy committed:
Add quay.io url in nfw

Co-authored-by: Hervé Quatremain <herve4m@users.noreply.github.com>

niroy commented:
> You also need to connect the firewall. I think you need to add another private subnet that hosts the firewall (between rEgressVpcPrivateSubnet and rEgressVpcPublicSubnet), add a route table, and update the other route tables. Example: https://medium.com/slalom-technology/secure-internet-access-egress-filtering-with-aws-network-firewall-ddf52ae121f9

Yes, for this one I found this guide: https://docs.aws.amazon.com/whitepapers/latest/building-scalable-secure-multi-vpc-network-infrastructure/using-nat-gateway-with-firewall.html
While, this perfectly fits our requirement, I'm still trying to figure out the routing table updates needed. Mainly confusion is on Transit gateway to egress vpc route and the blackhole route. I think we should discuss some more on this before we proceed

herve4m commented:
> 
 I've created a diagram with the current (non-working) template and one possible solution:

https://drive.google.com/file/d/18HfyFNGH3IU43SC4I6J4tltdjUBakz8R/view

We can use it to discuss and comment.

You're right, it's tricky if we go with blackhole routes (but that increases security between connected VPCs). Let's research on that.

herve4m committed:
Fixing the TransitGateway routes and cleaning up

herve4m committed:
Merge branch 'main' into niroy-pl-nfw

herve4m committed:
Removing "quay.io" from the domain list
".quay.io" also matches "quay.io"
https://docs.aws.amazon.com/network-firewall/latest/developerguide/stateful-rule-groups-domain-names.html

niroy committed:
Add firewall subnet and update routes

1. Added firewall subnet
2. Move the network firewall to reside on the firewall subnet
3. Update route tables for the three subnets in egressVPC
4. Changed the default subnet cidr value for the egressVPC subnets

niroy committed:
Fix course code

had changed course code to save delete stack time while testing

niroy committed:
Update network firewall rule to allow access.redhat.com

This is needed to mitigate the below error during rosa cluster creation

Name: 'egress', State: 'failed', Documentation reference: 'https://docs.openshift.com/rosa/rosa_planning/rosa-aws-prereqs.html#osd-aws-privatelink-firewall-prerequisites', subnet-019f2b0af334034f7: 'egressURL error: registry.redhat.io:443', subnet-019f2b0af334034f7: 'egressURL error: registry.access.redhat.com:443'

herve4m committed:
Fixing pre-commit YAML lint

herve4m committed:
Organizing the project files
